src/Common/src/System/Collections/Generic/EnumerableHelpers.Linq.cs

```diff
@@ -0,0 +1,61 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Collections.Generic
+{
+    /// <summary>Internal helper functions for working with enumerables.</summary>
+    internal static partial class EnumerableHelpers
+    {
+        /// <summary>Gets the count of the specified enumerable.</summary>
+        /// <param name="source">The enumerable to count.</param>
+        /// <returns>The count of the enumerable.</returns>
+        internal static int Count<T>(IEnumerable<T> source)
+        {
+            Debug.Assert(source != null);
+
+            var collectionOfT = source as ICollection<T>;
+            if (collectionOfT != null)
+            {
+                return collectionOfT.Count;
+            }
+            
+            var collection = source as ICollection;
+            if (collection != null)
+            {
+                return collection.Count;
+            }
+            
+            IEnumerator<T> enumerator = source.GetEnumerator();
+            return CountAndDispose(enumerator, enumerator);
+        }
+
+        /// <summary>
+        /// Gets the number of items in the enumerator, and then disposes it.
+        /// </summary>
+        /// <param name="enumerator">The enumerator to count.</param>
+        /// <param name="disposable">The enumerator typed as an <see cref="IDisposable"/>.</param>
+        /// <returns>The count of the enumerator.</returns>
+        /// <remarks>This method should be kept non-generic to avoid generated code bloat.</remarks>
```

Comment:
@stephentoub 

> This only ends up saving on generated assembly if this code is never hit for a given value type T (or any reference type) but the earlier part of the method is.

Actually, I believe this saves generated assembly in 2 scenarios:

1. Like you mentioned, if for a particular value type T the collection always implements one of the interfaces and does not hit this codepath, then this will save.

2. Another scenario is if this codepath is hit twice by two different value types, it will only be generated once, since it's non-generic. Example

```cs
lazyBytes.Count<byte>();
lazyInts.Count<int>();
```

> Do you have example libraries or apps in mind,

@stephentoub, I have been trying to measure the impact this has on Roslyn for a couple of days. However, I'm running into some runtime issues using Roslyn with a custom build of coreclr. If you'd like, I can remove this part of the PR out for now, and make a new PR when I get numbers.